Handle multipart content

annapowellsmith opened the pull request:
Please don't review this unless you are @benilovj!

Add info pages for multi-part content like detailed guides. 

These have lead metrics and detailed per-page metrics.

benilovj commented on app/views/info/_per_page_metrics.erb:
```diff
@@ -0,0 +1,27 @@
+      <tbody>
+        <% per_page_metrics.each do |t, a| %>
+          <tr>
+          <td class='page'><a href="<%= t %>"><%= t %></a></td>
```
it'd be nice to have the page title here as the link text, rather than the URL itself

benilovj replied:
I suppose that the page title would need to be exposed in the `metadata-api first`...

annapowellsmith replied:
Agreed. Will add.

annapowellsmith replied:
Now added to metadata-api: https://github.com/alphagov/metadata-api/pull/10

benilovj commented on app/controllers/info_controller.rb:
```diff
@@ -13,8 +14,17 @@ def show
       end
-      @lead_metrics = lead_metrics_from(@artefact, metadata.fetch("performance"))
+      part_urls = []
+      details = @artefact.fetch("details")
```
This method has now got a lot going on... any way to slim it down some?

annapowellsmith replied:
Done!

benilovj commented:
Unless I'm misreading the logic, I think the metrics for the base slug might be getting lost in the per-page calculations ie the data for `/housing-benefit` is getting thrown away, even though it's effectively the same content as `/housing-benefit/overview`. I think this can be fixed by treating the base slug as a `part_url`.

benilovj commented on spec/support/metadata_api_helpers.rb:
```diff
@@ -97,6 +110,39 @@ def metadata_api_response_for_multipart_artefact
       ]
+      response["performance"]["page_views"] += [
+        {"value"=>25000,
+         "path"=>"/apply-uk-visa/part-1",
```
to be closer to a real example, it's worth adding some example data for `"path"=>"/apply-uk-visa"` here, and the same applies further down for `response["performance"]["searches"]` and `response["performance"]["problem_reports"]`

annapowellsmith replied:
Good spot, I've amended this as you suggest.

annapowellsmith commented:
We're going to need some wording changes - e.g. "Searches started from this page" doesn't really work for multipart guides, and I think it would be good to clarify that the pageviews are averaged across all pages. But I'll make these requests separately.

benilovj commented:
I suspect that we've not got the lead metrics calculations right. It appears that they're being averaged across every part url and every day, where as they should probably be summed across part urls, and **then** averaged across the days:

![image](https://cloud.githubusercontent.com/assets/23801/5865436/329ccd36-a281-11e4-833b-2861c0bb75ae.png)

ie I think the lead pageviews metric should equal the sum of the part averages

annapowellsmith commented:
Yes, I think initially we decided to take the average of the averages, but on reflection I agree that the sum is better. 

(This will also make multipart guides more consistent with smart answers, because GA doesn't count page views separately for individual URLs in smart answers, so we'll have to present summed page views for smart answers whether we like it or not.) 

In that case we should sum problem reports and searches too. I'll get on it!

annapowellsmith commented:
...though this does mean it's much more difficult to put these numbers in context. 

I guess we need to use the sum-across-pages for the headline metrics, and then the average-across-all-pages in order to compare with other content.

benilovj commented:
> I guess we need to use the sum-across-pages for the headline metrics, and then the average-across-all-pages in order to compare with other content.

I think sum-across-pages makes more sense for the headline metrics. My understanding is that some guides are treated as a single entity that targets one high-level user need (so the sums are important for this(, whereas other guides target individual user needs (the individual page metrics are more useful in those cases).

I'd defer the metrics that we need for comparison until later - I don't feel like we have done enough user research to understand what would be useful in that context.

annapowellsmith commented:
Thanks Jake. I've added sum-across-pages in a new commit, plus some other minor things. I've tried to make naming as intuitive as possible for sum-across-pages, and write helpful commit messages, but will be interested to hear your feedback.

I'm hoping to add some basic comparison metrics (probably not for multi-part guides) early next week. I'll user test them on unsuspecting content designers before committing anything!

benilovj commented:
> I'm hoping to add some basic comparison metrics (probably not for multi-part guides) early next week. I'll user test them on unsuspecting content designers before committing anything!

:+1: :+1: :+1:

benilovj commented on app/views/info/_lead_metrics.html.erb:
```diff
@@ -1,16 +1,33 @@
 <section id="lead-metrics">
   <% if lead_metrics %>
+    <%- if format == 'guide' %>
```
I'm not keen on hard-coding the format here - `guide` isn't the only format that's multi-part (`programme` is also a multi-part format, see [an example here](https://www.preview.alphagov.co.uk/1619-bursary-fund)) and there could be other new formats introduced. I'd prefer a duck-typing approach (anything that's multi-part).

benilovj replied:
I think the confusion that I went through could be avoided by having a boolean `is_multipart` flag instead of a `format` variable?

annapowellsmith replied:
I did it this way to future-proof it for smart answers, I think we're going
to need different wording in the view for them.

They're different because they don't have unique page views for each URL,
due to the way GA records its metrics, but they do have problem reports.
On 24 Jan 2015 21:42, "Jake Benilov" <notifications@github.com> wrote:

> In app/views/info/_lead_metrics.html.erb
> <https://github.com/alphagov/info-frontend/pull/38#discussion_r23499635>:
>
> > @@ -1,16 +1,33 @@
> >  <section id="lead-metrics">
> >    <% if lead_metrics %>
> > +    <%- if format == 'guide' %>
>
> I think the confusion that I went through could be avoided by having a
> boolean is_multipart flag instead of a format variable?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/alphagov/info-frontend/pull/38/files#r23499635>.
>

benilovj commented on app/views/info/_lead_metrics.html.erb:
```diff
@@ -1,16 +1,33 @@
 <section id="lead-metrics">
   <% if lead_metrics %>
+    <%- if format == 'guide' %>
+    <h2>Metrics for entire guide</h2>
```
This could be made more generic with `<h2>Metrics for entire <%= format%></h2>` or `Metrics across all parts`?

benilovj commented on lib/performance_data/metrics.rb:
```diff
@@ -16,6 +16,18 @@ def problem_reports_weekly_average
       average(@data[:problem_reports]) * 7
     end
 
+    def unique_pageviews_multipart_average
```
I think there's another class trying to get out of this one (something like `MultiPartMetrics` maybe?). It could have the same method signature, and that'd save on a lot of ugly conditional logic in the views.

annapowellsmith commented:
In the latest commits, I've added a new `MultiPartMetrics` class and I've simplified the template logic down to a boolean `is_multipart`. On reflection I don't think we need different wording for different formats. 

The new class is a huge improvement, thanks for suggesting that. 

In the final commit I've also added support for smart answers, contingent on https://github.com/alphagov/metadata-api/pull/11

benilovj commented:
This looks great! :+1:

annapowellsmith commented:
Hooray!
Thanks for reviewing :)
